src/JsonRpc/InputHandler.cs

```diff
@@ -39,113 +58,238 @@ public class InputHandler : IInputHandler
             ILoggerFactory loggerFactory,
             ISerializer serializer,
             int? concurrency
-            )
+        )
         {
-            if (!input.CanRead) throw new ArgumentException($"must provide a readable stream for {nameof(input)}", nameof(input));
-            _input = input;
+            _pipeReader = pipeReader;
             _outputHandler = outputHandler;
             _receiver = receiver;
             _requestProcessIdentifier = requestProcessIdentifier;
             _requestRouter = requestRouter;
             _responseRouter = responseRouter;
             _serializer = serializer;
             _logger = loggerFactory.CreateLogger<InputHandler>();
-            _scheduler = new ProcessScheduler(loggerFactory, concurrency);
-            _inputThread = new Thread(ProcessInputStream) { IsBackground = true, Name = "ProcessInputStream" };
+            _scheduler = new ProcessScheduler(loggerFactory, concurrency,
+                new EventLoopScheduler(_ => new Thread(_) {IsBackground = true, Name = "InputHandler"}));
+            _headersBuffer = new Memory<byte>(new byte[HeadersFinishedLength]);
+            _contentLengthBuffer = new Memory<byte>(new byte[ContentLengthLength]);
+            _contentLengthValueBuffer = new byte[20]; // Max string length of the long value
+            _contentLengthValueMemory =
+                new Memory<byte>(_contentLengthValueBuffer); // Max string length of the long value
+            _stopProcessing = new CancellationTokenSource();
+
+
+            _disposable = new CompositeDisposable {
+                Disposable.Create(() => _stopProcessing.Cancel()),
+                _stopProcessing,
+                Disposable.Create(() => _pipeReader?.Complete()),
+                _scheduler,
+            };
         }
 
         public void Start()
         {
-            _scheduler.Start();
-            _outputHandler.Start();
-            _inputThread.Start();
+            ProcessInputStream(_stopProcessing.Token).ContinueWith(x => {
+                if (x.IsFaulted) _logger.LogCritical(x.Exception, "unhandled exception");
+            });
         }
 
-        // don't be async: We already allocated a seperate thread for this.
-        private void ProcessInputStream()
+        private bool TryParseHeaders(ref ReadOnlySequence<byte> buffer, out ReadOnlySequence<byte> line)
         {
-            // some time to attach a debugger
-            // System.Threading.Thread.Sleep(TimeSpan.FromSeconds(5));
+            // TODO: This might be simplified with SequenceReader...
+            // Not sure we can move to purely .netstandard 2.1
```

Comment:
please don't move to netstandard2.1 since it doesn't work in .NET Framework :(

Reply:
Yeah, that's an old comment, this PR works on netframework, netcoreapp2.1 and netcoreapp3.1 and support will be for the foreseeable future.

There are just things thank netstandard2.1 makes easier.... either way I learned a heck of lot by not having it 😄